raw-spans-grouper/src/main/java/org/hypertrace/core/rawspansgrouper/TraceEmitPunctuator.java

```diff
@@ -0,0 +1,137 @@
+  private ProcessorContext context;
+  private KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore;
+  private KeyValueStore<String, Long> traceEmitTriggerStore;
+  private To outputTopicProducer;
+  private long groupingWindowTimeoutMs;
+  private Cancellable cancellable;
+
+  public TraceEmitPunctuator(String key, ProcessorContext context,
+      KeyValueStore<String, ValueAndTimestamp<RawSpans>> inflightTraceStore,
+      KeyValueStore<String, Long> traceEmitTriggerStore,
+      To outputTopicProducer,
+      long groupingWindowTimeoutMs) {
+    this.key = key;
+    this.context = context;
+    this.inflightTraceStore = inflightTraceStore;
+    this.traceEmitTriggerStore = traceEmitTriggerStore;
+    this.outputTopicProducer = outputTopicProducer;
+    this.groupingWindowTimeoutMs = groupingWindowTimeoutMs;
+  }
+
+  public void setCancellable(Cancellable cancellable) {
+    this.cancellable = cancellable;
+  }
+
+  /**
+   * @param timestamp correspond to current system time
+   */
+  @Override
+  public void punctuate(long timestamp) {
+    // always cancel the punctuator else it will get re-scheduled automatically
+    cancellable.cancel();
+
+    Long emitTs = traceEmitTriggerStore.delete(key);
+    if (emitTs <= timestamp) {
+      // Implies that no new spans for the trace have arrived within the last 'sessionTimeoutMs' interval
+      // so the trace can be finalized and emitted
+      ValueAndTimestamp<RawSpans> agg = inflightTraceStore.delete(key);
+      ByteBuffer traceId = null;
+      String customerId = null;
+      if (!agg.value().getRawSpans().isEmpty()) {
+        RawSpan firstSpan = agg.value().getRawSpans().get(0);
+        traceId = firstSpan.getTraceId();
+        customerId = firstSpan.getCustomerId();
+      }
+      List<RawSpan> rawSpanList = new ArrayList<>();
+      for (IndexedRecord r : agg.value().getRawSpans()) {
+        rawSpanList.add((RawSpan) r);
+      }
+
+      Timestamps timestamps = trackEndToEndLatencyTimestamps(timestamp, agg);
+      StructuredTrace trace = StructuredTraceBuilder
+          .buildStructuredTraceFromRawSpans(rawSpanList, traceId, customerId,
+              timestamps);
+
+      logger.debug("Emit trace_id=[{}], spans_count=[{}]", key, rawSpanList.size());
+      context.forward(null, trace, outputTopicProducer);
+    } else {
+      // implies spans for the trace have arrived within the last 'sessionTimeoutMs' interval
+      // so the session inactivity window is extended from the last timestamp
+      if (logger.isDebugEnabled()) {
+        logger.debug("Re-scheduling emit trigger for trace_id=[{}] to [{}]", key,
+            Instant.ofEpochMilli(emitTs + groupingWindowTimeoutMs));
+      }
+      long newEmitTs = emitTs + groupingWindowTimeoutMs;
+      traceEmitTriggerStore.put(key, newEmitTs);
+      // if current timestamp is ahead of newEmitTs then just add a grace of 1ms and fire it
+      long duration = Math.max(1, newEmitTs - timestamp);
```

Comment:
Should we keep it 10/100ms in this case?
- it may possible that 1ms can cause missed puncuator, so can be trigger 1ms, 2ms, 3ms, ...
- we may end-up-doing same processing in succession?

Reply:
made it 100ms